subsys/bluetooth/mesh/delayable_msg.c

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <zephyr/sys/slist.h>
+
+#include <zephyr/net/buf.h>
+#include <zephyr/bluetooth/mesh.h>
+
+#include "msg.h"
+#include "access.h"
+
+#define LOG_LEVEL CONFIG_BT_MESH_ACCESS_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_mesh_delayable_msg);
+
+static void delayable_msg_handler(struct k_work *w);
+static bool push_msg_from_delayable_msgs(void);
+
+struct delayable_msg_chunk {
+	sys_snode_t node;
+	uint8_t data[CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_CHUNK_SIZE];
+};
+
+struct delayable_msg_ctx {
+	sys_snode_t node;
+	sys_slist_t chunks;
+	struct bt_mesh_msg_ctx ctx;
+	uint16_t src_addr;
+	const struct bt_mesh_send_cb *cb;
+	void *cb_data;
+	uint32_t fired_time;
+	size_t len;
+};
+
+static struct {
+	sys_slist_t busy_ctx;
+	struct k_work_delayable random_delay;
+} access_delayable_msg = {.random_delay = Z_WORK_DELAYABLE_INITIALIZER(delayable_msg_handler)};
+
+K_MEM_SLAB_DEFINE_STATIC(delayable_msg_chunks, sizeof(struct delayable_msg_chunk),
+			 CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_CHUNK_COUNT, 4);
+
+K_MEM_SLAB_DEFINE_STATIC(delayable_msgs_ctx, sizeof(struct delayable_msg_ctx),
+			 CONFIG_BT_MESH_ACCESS_DELAYABLE_MSG_COUNT, 4);
+
+static void put_ctx_to_busy_list(struct delayable_msg_ctx *ctx)
+{
+	struct delayable_msg_ctx *curr_ctx;
+	sys_slist_t *list = &access_delayable_msg.busy_ctx;
+	sys_snode_t *head = sys_slist_peek_head(list);
+	sys_snode_t *curr = head;
+	sys_snode_t *prev = curr;
+
+	if (!head) {
+		sys_slist_append(list, &ctx->node);
+		return;
+	}
+
+	do {
+		curr_ctx = CONTAINER_OF(curr, struct delayable_msg_ctx, node);
+		if (ctx->fired_time < curr_ctx->fired_time) {
+			if (curr == head) {
+				sys_slist_prepend(list, &ctx->node);
+			} else {
+				sys_slist_insert(list, prev, &ctx->node);
+			}
+			return;
+		}
+		prev = curr;
+	} while ((curr = sys_slist_peek_next(curr)));
+
+	sys_slist_append(list, &ctx->node);
+}
+
+static struct delayable_msg_ctx *get_pending_msg(void)
+{
+	struct delayable_msg_ctx *pending_msg = NULL;
+	sys_snode_t *node = sys_slist_peek_head(&access_delayable_msg.busy_ctx);
+
+	if (node) {
+		pending_msg = CONTAINER_OF(node, struct delayable_msg_ctx, node);
+	}
+
+	return pending_msg;
+}
+
+static void reschedule_delayable_msg(struct delayable_msg_ctx *msg)
+{
+	uint32_t curr_time;
+	k_timeout_t delay = K_NO_WAIT;
+	struct delayable_msg_ctx *pending_msg = NULL;
+
+	if (msg) {
+		put_ctx_to_busy_list(msg);
+	}
+
+	pending_msg = get_pending_msg();
+
+	if (!pending_msg) {
+		return;
+	}
+
+	curr_time = k_uptime_get_32();
+	if (curr_time < pending_msg->fired_time) {
+		delay = K_MSEC(pending_msg->fired_time - curr_time);
+	}
+
+	k_work_reschedule(&access_delayable_msg.random_delay, delay);
```

Comment:
Can we avoid rescheduling if head didn't change? E.g. check if `get_pending_msg` before `put_ctx_to_busy_list` and after returns different message.

Reply:
I tried but it doesn't work. Being the head of the sorted context list doesn't mean being scheduled. For example, when a message is sent context is removed from the list. Here head of the list will be the same before and after `put_ctx_to_busy_list` but the delayed timer is not started at all.